### ISSUE TITLE
Add support for browser tab scope

### ISSUE THREAD

**Issue by mvysny** [vaadin/flow#13468]
### Describe your motivation

I'm building an app which performs bookings; the user can open the app in multiple tabs and launch a booking wizard in every tab. Each wizard instance may have a different booking holder configured, and this booking holder needs to be preserved even when the wizard is navigated away, or when the tab is reloaded via F5.

### Describe the solution you'd like

I need to bind certain information (a bean or a string) to a Browser Tab "scope" (using the word "scope" loosely here since I'm not using Spring). The information needs to survive F5 and all sorts of navigation: via UI.navigate(), via the back/forward button and via the user typing a new URL manually into the browser bar. The bean doesn't need to survive session close.

Having an API from Vaadin similar to `BrowserTab.getCurrent().setAttribute("foo", "bar")` would be excellent. The values would survive all sorts of navigations and also F5 page reload. When the tab is closed, the values would be removed and garbage-collected, either immediately or at unspecified time in the future. When the session is closed, the values would be removed and garbage-collected, either immediately or at unspecified time in the future.

Alternatively, the UI object could be able to provide a tab ID which would persevere even if UI is destroyed and re-created; that would allow me to create a HashMap from tab ID to my object in VaadinSession. The UI.getId() is null and UI.getUiId() increases with new UI instances and therefore can not be used as tab identifiers.

### Describe alternatives you've considered

* I tried to tie the information via `ComponentUtil.setData(UI.getCurrent())` but this doesn't survive F5.
* I tried having a `@PreserveOnRefresh`-annotated parent layout which does survive F5 but it doesn't survive browser back/forward button nor user typing a new URL manually.
* There's no tab identifier, so I can't create a HashMap from tab ID to my object in VaadinSession.

The abovementioned can be replicated with the following example app, just paste these classes into Skeleton Starter for Vaadin 14.8:

```java
@PreserveOnRefresh
public class MyLayout extends VerticalLayout implements RouterLayout {
    public MyLayout() {
        System.out.println("CONSTRUCTOR CALLED");
    }
}

@Route(value = "", layout = MyLayout.class)
public class MainView extends VerticalLayout implements HasUrlParameter<String> {
    public MainView() {
        add(new Button("Navigate - doesn't re-create MyLayout", e -> UI.getCurrent().navigate(MainView.class, "foobar")));
    }

    @Override
    public void setParameter(BeforeEvent event, @OptionalParameter String parameter) {
        System.out.println("Called with parameter " + parameter);
        System.out.println(UI.getCurrent());
        System.out.println(VaadinSession.getCurrent());
    }
}
```

The test scenarios:
1. Run the app, click the "Navigate - doesn't re-create..." Button and go back: both the UI and the MyLayout is re-created, so I can't tie my bean as a data to those things.
2. Run the app, click the "Navigate - doesn't re-create..." Button and go manually to a different URL, say http://localhost:8080/baz . Both the UI and MyLaoyut are re-created.

**Comment by Artur-** [comment]
At least earlier Vaadin used to set `window.name` to a generated string so that the server could find old relevant instances. This should stay the same also on back-forward but maybe `@PreserveOnRefresh` does not use this. The name even indicates that it is only for refresh

**Comment by mvysny** [comment]
Yeah, on client-side there's `window.name` with the value of "ROOT-2521314-0.6047083771634619" (and the value is preserved on page reloads and all navigations), but how can I transfer that value to server-side before views/layouts are instantiated?

**Comment by mvysny** [comment]
I don't think there's any other way, judging from the fact that vaadin itself uses window.name to preserve extended client details

**Comment by archiecobbs** [comment]
Is there a reason there can't be a method `Page.getId()` which returns on the server the same thing as `window.name` on the client?

Otherwise it appears impossible in Vaadin to distinguish browser tabs.

Related to that, it's impossible to know whether to preserve or clean up state when `onDetach()` is invoked in a scenario where you want to preserve state on a reload but also cleanup state on a "real" close.

Here's why: I've observed the following behavior when reloading tabs:

* Without `@PreservOnRefresh`:
  * Component `onDetach()`
  * New view `beforeEnter()` (`isRefreshEvent()` = false)
  * Component `onAttach()` - with a new `UI` instance
* With `@PreservOnRefresh`:
  * Component `onDetach()`
  * Same view `beforeEnter()` (`isRefreshEvent()` = true)
  * Component `onAttach()` - _also_ with a new `UI` instance

Side note: You might expect in the `@PreservOnRefresh` case that the same `UI` would be used on re-attach, but [that changed in Vaadin 10](https://vaadin.com/docs/latest/advanced/preserving-state-on-refresh#preserving-the-state-of-a-component).

So you can see the `onDetach()` is invoked (first) the same way in both cases. For a proper preserve/cleanup lifecycle your state would need to live in the view (which it normally does) and for that the only reliable "end of lifecycle" event is `onDetach()` (you don't get `beforeLeave()` events when tabs are closed) but if you use that then you can't preserve state on reloads, at least not without some hacks like "wait a few seconds to see if you get another `beforeEvent()` before cleaning up your state", etc. Even trickier would be preserving state when someone closes and the re-opens the same tab later (must be enough later that the keep-alive has timed out) - a solution for both would be remembering "unattached" state in the session and then when views are loaded finding and "attaching" the state back to the view.

**Comment by mvysny** [comment]
> Is there a reason there can't be a method Page.getId() which returns on the server the same thing as window.name on the client?

This would be amazing to have! This would allow us to implement tab scope rather easily. The `getId()` must not be asynchronous though - we already have that, the `ClientExtendedDetails.getWindowName()`.

Alternatively, if there would be some support for some kind of "continuity" - Vaadin would let me know that the current UI is closing but is replaced by this new UI - that would also solve this issue since we could easily transfer the tab-scoped values to the new UI.

However, I found out that using `@RouteScopeOwner(MainLayout.class)` on a route actually scopes the route to a tab, so this might be useful if you're using Spring. I've updated https://mvysny.github.io/vaadin-ui-scope/ to reflect this.

**Comment by mvysny** [comment]
In short, in Spring-Boot-based projects you can add `@RouteScopeOwner(MainLayout.class) @RouteScope @SpringComponent` to your routes and beans to have them route-scoped. However, I propose to have this ticket open since it would be great to have an API that works without Spring.

API proposal: `UI.getTabScope()` would return `TabScopedAttributes` which would have API similar to `Session.getAttribute()`/`Session.setAttribute()` and would be tab-scoped. Alternatively `TabScopedAttributes.getCurrent()` would return the currently scoped attributes. Additionally a `VaadinSession` could allow you to list all active tab scopes.

**Comment by TatuLund** [comment]
> API proposal: UI.getTabScope() would return TabScopedAttributes which would have API similar to

There is already

`ComponentUtil.setData(ui, "attributeName", value);`

Can be used with any component, so UI, MainLayout, or what ever.

**Comment by mvysny** [comment]
The problem is that UI is never tab-scoped since it never survives F5. Any data tied to the UI instance is gone after you press F5.

**Comment by TatuLund** [comment]
> The problem is that UI is never tab-scoped since it never survives F5. Any data tied to the UI instance is gone after you press F5.

Can be used with any component, so UI, **MainLayout**, or what ever. MainLayout survives F5.

**Comment by mvysny** [comment]
True, however there is another problem. Say we have `MainLayout` annotated with `@RouteScopeOwner(MainLayout.class) @RouteScope`. The problem is that the `MainLayout` instance isn't available in early phases:

* In the `UIInitListener`
* In the route's constructor - at that point nothing is connected to the current UI so you can't walk UI children to find `MainLayout`.

Sure, in certain cases you can work around this by postponing the lookup to route's `onAttach()` function, but in certain cases it may not be possible: for example when migrating an unfamiliar Vaadin 8 project which expects the tab scope to be available very early on, in Vaadin8's `UI.init()` function.

**Comment by mvysny** [comment]
I've implemented an example app (pure servlet, no Spring) which demoes browser tab scope with:
* An init listener that's called exactly once per tab scope, and before all routes/layouts are instantiated
* A tab scoped values and routes
* Works even without `@PreserveOnRefresh`
* Both Vaadin 24 and Vaadin 23 examples are available

https://github.com/mvysny/vaadin-tab-scope-example

**Comment by TatuLund** [comment]
I think there is a lot of detail here in the discussion that is not fitting under the original title. The Browser Tab Scope do exists in Vaadin 23/24 in form of RouteScope. So that is a done deal. What here are being discussed in addition that is requirement for some sort of life cycle callbacks that would be more analogous to how Vaadin 8 worked. And as you are not supposed to extend UI in Vaadin 23/24 that will be pretty hard to make 100% same.

There is already some equivalents, e.g. on MainLayout there are five options to use 1) constructor 2) onAttach 3) beforeEnter 4) afterNavigation 5) constructor / post construct of the injected RouteScope bean. 

IMHO, this ticket should be closed and a new one opened with the topic of callbacks what ever it may be and add analysis of the use cases where neither of the existing four do not work.

**Comment by mvysny** [comment]
Actually, let's keep this ticket open. The reason is that:

- Tab scope should survive bookmarkable URLs/`getDriver.get()` if that's possible within given browser. The guarantees of `@PreserveOnRefresh` are weak since the annotation only promises to keep the instances on reload, but not on navigation. See https://github.com/vaadin/flow/issues/21141#issuecomment-2733321079
- By extension, Route-scoping to the `@PreserveOnRefresh`-annotated layout doesn't provide necessary guarantees.

A new, clean API should therefore be created, which would:

- promise to do its best to preserve tab-scoped values as much as possible (including navigation, both from within the app, by clicking a bookmarked URL and driver.get()), except cases where that's impossible (window.name is changed by the browser).
- Introduce a clean initialization chain with strong promises: e.g. the init listener is fired once before any route/layout is instantiated. The solutions mentioned by @TatuLund don't hold such strong promises. Let's reiterate:

> on MainLayout there are five options to use 1) constructor 2) onAttach 3) beforeEnter 4) afterNavigation 5) constructor / post construct of the injected RouteScope bean.

1. Constructor of the layout can be run after the view's constructor itself has been called.
2. onAttach is run after the routes/layouts are instantiated. Depending on the app, it might be too late to perform initialization.
3. beforeEnter - same thing as above.
4. afterNavigation - same as above
5. constructor / post construct of the injected RouteScope bean - this has a couple of drawbacks: the bean must be injected otherwise it is not instantiated eagerly; the time of instantiation depends on whether it's a counstructor-based injection or field-based injection, and again no guarantees are provided that this runs before any route/layout is instantiated.

**Comment by archiecobbs** [comment]
> Another alternative is to say that it's "tab scoped" within the confines of how each browser choses to define when one "logical tab" (i.e. `window.name`) ends and a new one is created within the same "visual tab". This might be sufficient since there will anyways also be other cases that the end user perceives as continuing in a previous context even though they are actually closing the tab and opening a new one.

This seems OK to me - to reiterate, my primary concern is not user friendliness but rather proper lifecycle management of server-side resources (see previous comment "it's impossible to know whether to preserve or clean up state..."). IMHO proper management of resources on the server is vastly higher priority than user friendliness. If some kind of workaround is inevitable, it should live at the user friendliness layer, not the layer that seeks to preserve server-side integrity. Put another way, screwing up your server-side state is terrible for user friendliness!

**Comment by mvysny** [comment]
My primary concern is to have a viable replacement for Vaadin 8 UI scope. If the user closed a tab and opened a new one, Vaadin 8 UI wouldn't survive that, so for me it's acceptable to lose the tab scope too.

**Comment by mvysny** [comment]
> Is there any other technical difference aside from #21139

Not that I'm aware of. Once #21139 is fixed, I'm happy. The problem is that #21139 can be rejected and closed as won't fix, since technically it works according to the spec. If that will be the case, we definitely need a proper tab scope.

**Comment by Legioth** [comment]
I realized there's one fundamental challenge with a tab scope and that's how to know when to close the scope. The problem is that the browser doesn't let JavaScript know about what happens to a tab, but only to individual documents opened inside that tab.

When an event like `unload` is fired, it might be because the user closed the tab, or because they reloaded the page, or because they clicked a link that does a regular navigation. For a reload, a new document will be loaded within roughly one round trip delay whereas for navigation, the user might navigate with the back button 1 minute or 1 day later. Similarly for closed tabs, most browsers have a "reopen last closed tab" action.

This is similar to UI instances for which we originally implemented a heartbeat mechanism to be able to detect UIs that no longer seem to be in use. The same mechanism could indeed also be used to keep tab scopes alive based on the most recent heartbeat from any UI inside each tab. But one significant difference is that we can in many cases eagerly detect closed UIs and do cleanup e.g. based on those `close` events or whenever a new UI instance is boostrapped for the same tab. Those mechanisms would not be available for tab scopes which means that the only thing that remains is heartbeats and session timeout.

**Comment by mvysny** [comment]
With sessions, we're used to the following behavior: Once user closes the last tab and the session is eventually destroyed, the user gets a new session on browser tab/window reopen. That is a reasonable expectation since one can't expect the session to remain around indefinitely, since that would be a very definition of a memory leak.

I think the same way of thinking also applies to tabs:
- Once a tab has been closed for "some time" (configurable), the scope is gone and destroyed.
- Once a session is closed, all tab scopes are closed too.

I'd say the reasonable value for "some time" is the same value used to timeout unresponsive UIs: three times the UI heartbeat duration. Once the tab misses three UI heartbeats, it is eligible for cleanup.

Note that we shouldn't cleanup the scope once all of associated UIs are closed: there's a brief window upon tab reload when the old UI is closed but a new one is not created yet. The timeout above should help in this scenario: during the page reload, the UIs are all closed but there will be a fairly recent heartbeat, and therefore the tab scope won't get closed.
